[STEP 1 & STEP 2] volga, summer

shinhee-rebecca opened the pull request:
@wonhee009

STEP1
- [x] 메뉴 출력 및 메뉴의 잘못된 입력 처리
- [x] 학생추가
- [x] 종료
- [x] 학생 삭제

STEP2
- [x] STEP1 리뷰사항 반영
- [x] 성적 추가
- [x] 성적 삭제
- [x] 평점보기

질문 사항
1. Student 구조체를 클래스로 변경하였습니다. 그 이유는 Student 원본의 내부를 변경할 필요가 있어서였습니다. Student를 구조체로 사용하면서 원본의 내부를 변경할 수 있는 방법이 있을까요?
2. 하나의 객체가 하나의 역할을 수행할 수 있도록 나름대로 노력을 해봤습니다. 개선할 수 있도록 조언을 부탁드립니다.

고려 사항
1. 객체나 변수의 이름을 지을 때 자신의 역할과 책임이 분명하게 드러나도록 짓기 위해 노력했습니다.
2. 하나의 객체가 하나의 역할을 수행할 수 있도록 노력했습니다.

Commit:
chore: unit test를 위한 타겟 생성

Commit:
test: input에 대한 test case 추가

volgaHwang committed:
feat: InputError 구현

volgaHwang committed:
feat: GradeManager 구조체 구현 반영

- 내부에 InfoMessage를 작성해서 guide를 따로 분리했습니다.
- menuChoice 함수를 사용해서 <String, InputError> 에 대해 에러처리를 할수 있도록 구현했습니다.
- menuChoice 내부에 switch 부분은 리팩토링 예정입니다.

Commit:
feat: GradeManager의 전체 실행 흐름 추가

- 입력을 돕는 inputManager 구현
- run 구현

Signed-off-by: kakao <kakao@summer-love-pc.local>

volgaHwang committed:
feat: StudentManager 추가구현

- GradeManager에서 StudentManager를 소유하게 했습니다.
- StudentManager는 add/delete 프로토콜을 통해 동작하게 됩니다.
- StudentManager에서 Student는 Name만으로 중복 검사를 진행하게 합니다.

Commit:
feat: addStudent, deleteStudent 구현

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,192 @@
+    }
+}
+
+struct GradeManager {
```
struct로 정의한 이유가 있을까요?

volgaHwang replied:
안녕하세요 라자냐

Struct를 사용한 이유는 다음과 같습니다.
일단, Class를 사용할지 Struct를 사용할지 Summer랑 페어프로그래밍 전에 논의를했는데요,

1. Apple Developer Academy에서 제시하는 Struct / Class를 사용하는 가이드 라인을 볼 경우
- Use structures by default.
- Use classes when you need Objective-C interoperability.
- Use classes when you need to control the identity of the data you’re modeling.
- Use structures along with protocols to adopt behavior by sharing implementations.

위의 글에서 제시된 것 처럼, 해당 집합체가 Objective-C와 interoperability가 필요없는 상태고
딱히 동일성(Identity)에 대해 신경을 쓸 이유가 없다고 생각했기 때문입니다.
단순하게 GradeManager로서 동작확인이 목적이라면 Default로 제시되어있는 Struct를 쓰는 것이 좋겠다고 판단했습니다.

2. 두번째 이유로는 굳이 가이드라인에서 제시하는 위의 네 문장을 제외하더라도 주소값을 통해 어떤 무언가를 해야 할 경우에 사용하게 될텐데, 해당 과제에서는 Reference Type을 사용할 이유가 없다고 판단했기 때문입니다.

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,192 @@
 import Foundation
+
+enum InputError: Error {
+    case null
```
사용자가 아무것도 입력하지 않았을때 에러인거로 보이는데요.
`null`이라고 케이스 네이밍한 이유가 있을까요?
swift에서는 `null`이 아닌 `nil`을 사용하기도 하고, `InputError.null`은 에러가 비었다는건지 빈 입력에 대한 에러인지 알기 어려울거 같습니다.
`null`보다는 `empty`같은 네이밍은 어떤가요?

wonhee009 replied:
또한 `wrong`이라는 네이밍보다는 구체적인 네이밍은 어떤가요?
해당 에러가 발생하는 상황을 보니 없는 메뉴에 대한 input이 들어왔을때 발생하는거 같은데요.
이런 상황이 잘 들어나는 네이밍이면 좋을거 같습니다.

volgaHwang replied:
null이라고 네이밍을 지은 이유는 비어있는 상태를 어떻게 표현하면 좋을지 생각하다가 null이 적당하겠구나! 하는 생각에 지었습니다.
말씀 주신대로 ```null```보다는 ```empty```가 좋을것 같고 정확하게 어떠한것이 ```empty```한것인지에 대해 명시적으로 알려줄수 있는 것이 있는지 리팩토링 해야할것같습니다.

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,192 @@
+        }
+        
+        switch choice {
+        case "1", "2", "3", "4", "5", "X":
```
"1", "2", .... 가 메뉴 문자열로 보입니다.
enum을 이용해 메뉴를 정의하고 enum을 활용해보는건 어떤가요?
``` swift
enum Menu: String {
   case .studentAdd = "1"
   ...
}

guard let choiceMenu = Menu(rawValue: choice) else {
   return .failure(.wrong)
}
...
```

이런식으로 enum과 rawValue를 이용한다면 문자열을 주고받지 않고 enum Type으로 주고 받을 수 있을거 같습니다.

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,192 @@
+}
+
+struct DefaultStudentManager: StudentManager {
+    enum InfoMessage: String {
+        case addSuccess = " 학생을 추가했습니다."
+        case addFailure = "입력이 잘못되었습니다. 다시 확인해주세요."
+        case deleteSuccess = "학생을 삭제하였습니다."
+        case deleteFailure = " 학생을 찾지 못했습니다."
```
프로그램에서 사용하는 문자열 상수를 정의한거로 이해했습니다.
rawValue를 사용한 이유가 있을까요?
``` swift
enum InfoMessage {
   static let addSuccess = "학생을 추가했습니다."
   ...
}
```
이런 식으로 타입 프로퍼티를 이용할 수 있습니다.
`InfoMessage`가 `DefaultStudentManager` 내에서만 사용되는거 같은데요.
private으로 제한할 필요는 없을까요?

volgaHwang replied:
- 프로그램에서 사용하는 문자열 상수를 정의한거로 이해했습니다.
rawValue를 사용한 이유가 있을까요?
    - rawValue를 통해 InfoMessage 의 printing을 사용하면, ```InfoMessage.add.printing()```같이 출력하는 방식을 선택했기 때문입니다.
- ```InfoMessage```가 ```DefaultStudentManager``` 내에서만 사용되는거 같은데요. private으로 제한할 필요는 없을까요?
    - 말씀 주신대로, 불필요하게 외부에 노출이 될 필요가 없는 것 같습니다.

wonhee009 replied:
`InfoMessage.add.printing()`과 같은 방식을 사용하려 해당 rawValue로 구현한거로 이해했습니다.
`print(InfoMessage.add.rawValue)`가 위의 코드와 동일하게 동작할거 같은데요.
다른 사람이 봤을때에는 `.printing()`, `print()` 어떤걸 써야할지 혼란스러울거 같기도 하고, 누군가 `printing` 메서드의 존재를 알려주지 않으면 혼용되서 사용할 가능성이 높아보입니다.
보통 이런 문자열 상수는 `static let`처럼 타입 프로퍼티를 이용해서 정의하는데요. 이런 방식은 어떨까요?

volgaHwang replied:
안녕하세요. 라자냐.

제가 지난 금요일에 다른 공적으로 처리해야할 문서들 때문에 참여를 못해서 늦게나마 답변을 답니다.
리뷰에 대한 이해를 제대로 하고있는게 맞는지 살짝은 저도 혼동이 되는데요 ㅎㅎㅎ

굳이 ```InfoMessage```라는 Enumerate에 String으로 ```rawValue```를 사용하는 것보다는
케이스에 맞게 ```static```을 이용한 상수로 출력을 해주는 방식을 이용하는 것을 추천하신다고 이해를 했습니다.

만약 제가 이해한 부분이 맞다면, 저도 이 부분에 대해서 동의합니다.
이유는 두 가지인데요, case에서 한번 더 해당 부분에 대한 ```rawValue```를 찾는 것이 조금은 비효율적인 코드로 보이고,
실제로 코드를 작성하는 경우에 있어서도 ```rawValue``` 키워드를 그대로 사용하는 것은 가독성이 좋지 않다고 판단이 들기 때문입니다.

그리고 말씀 주신 printing(), print() 키워드에 대해서도 헷갈리는 지점이 있다는 것에 동의합니다.
이 부분도 썸머와 함께 논의 해서 수정을 진행해보고 다시 PR을 올려보겠습니다 :)

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,192 @@
+    }
+}
+
+struct DefaultStudentManager: StudentManager {
```
`Default`는 어떤 의미로 붙은건가요?

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,192 @@
+    private var list = Set<Student>()
+    
+    func isDuplicate(name: String) -> Bool {
+        for student in list {
+            if student == name {
+                return true
+            }
+        }
+        return false
```
[Set](https://developer.apple.com/documentation/swift/set)의 `contains` 메서드를 활용할 수 있지 않을까요?

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,192 @@
+    }
+    
+    mutating func add(name: String) {
+        if isDuplicate(name: name) {
+            InfoMessage.addFailure.printing(name: name)
+            return
+        }
```
early exit 상황으로 보이는데요. [guard문](https://docs.swift.org/swift-book/LanguageGuide/ControlFlow.html) 활용해보는건 어떤가요?

wonhee009 commented on GradeManager/main.swift:
```diff
@@ -5,3 +5,192 @@
+    }
+    
+    mutating func run() {
+        while true {
```
while true보다는 프로퍼티를 통해 제어하는건 어떨까요?
``` swift
let canExit: Bool

while canExit { ... }
```
이런 식으로요. 현재는 어떤 상황에서 탈출하는지 로직을 보지 않고는 알기 힘든거 같습니다.

volgaHwang replied:
오.. 이 부분은 처음에 굉장히 고민을 많이했던 부분인데 한번 참조에서 리팩토링 진행해보겠습니다 :)

wonhee009 commented on UnitTestGradeManager/UnitTestGradeManager.swift:
```diff
@@ -0,0 +1,107 @@
+        }
+    }
+    
+    func test_input_menu_error() {
+        let result = sut.menuChoice()
+        switch result {
+        case .success:
+            XCTFail()
+        case .failure:
+            XCTAssert(true)
+        }
+    }
```
정의한 errorType (null, wrong)에 대해 케이스를 나눠서 테스트 코드를 추가해주세요.

wonhee009 commented on UnitTestGradeManager/UnitTestGradeManager.swift:
```diff
@@ -0,0 +1,107 @@
+        }
+    }
+
+    override func setUpWithError() throws {
+        // Put setup code here. This method is called before the invocation of each test method in the class.
+    }
+
+    override func tearDownWithError() throws {
+        // Put teardown code here. This method is called after the invocation of each test method in the class.
+    }
+
+    func testExample() throws {
+        // This is an example of a functional test case.
+        // Use XCTAssert and related functions to verify your tests produce the correct results.
+        // Any test you write for XCTest can be annotated as throws and async.
+        // Mark your test throws to produce an unexpected failure when your test encounters an uncaught error.
+        // Mark your test async to allow awaiting for asynchronous code to complete. Check the results with assertions afterwards.
+    }
+
+    func testPerformanceExample() throws {
+        // This is an example of a performance test case.
+        measure {
+            // Put the code you want to measure the time of here.
+        }
+    }
```
사용하지 않는 코드라면 삭제 부탁드립니다.

wonhee009 commented on UnitTestGradeManager/UnitTestGradeManager.swift:
```diff
@@ -0,0 +1,107 @@
+
+import XCTest
+
+class UnitTestGradeManager: XCTestCase {
```
addStudent, deleteStudent에 대한 테스트 코드는 필요 없나요?

volgaHwang replied:
테스트 코드는 필요하다고 생각합니다.

[f994026](https://github.com/2023newkrew/ios-grade-manager/pull/1/commits/f994026d2c31bc4f14d3d88397381a0077e9417d) 말씀 하신 테스트에 대한 패치를 전달드립니다!

wonhee009 reviewed (changes requested):
수고하셨습니다.
코멘트 남겨두었으니 확인 부탁드립니다.
(이후 PR에서는 어떤 부분을 고민했는지 PR 내용에 추가 부탁드립니다.)

volgaHwang committed:
refactor: input Manager 테스트 코드 수정

- GradeManager 에서 menuChoice 부분이 바뀜에 따라 테스트코드를 수정하여 반영합니다.

volgaHwang committed:
refactor: 불필요한 테스트코드 삭제

- 불필요한 테스트 코드를 삭제합니다.

volgaHwang committed:
refactor: GradeManager에서 Menu Enumerate를 정의했습니다.

- 기존에 매직넘버로 사용하던, 메뉴 번호 String 1,2,3,4,5,X에 대하여 enumerate로 Menu 타입을 정의하였습니다.
- 또한 기존에 inputError로 모호하게 사용하던 null, wrong Naming에 대하여 하나로 축약하여 invalidUserInput으로 처리하였습니다.
-

volgaHwang committed:
refactor: StudentManager에서 isDuplicate 메소드 변경사항 반영

- for문을 사용한 루프탐색에서 Set에서 메소드 특성을 이용한 contains로 변경하여 반영합니다.

volgaHwang committed:
refactor: run while loop 수정사항 반영

- canOperate 를 추가하여, menu가 .exitGradeManager일 경우 상태를 변경하여 프로그램 동작을 제어하게 수정했습니다.

volgaHwang committed:
refactor: 테스트 코드 리팩터 전에 테스트 케이스 수정했는데, 반영을 늦게 했습니다..

volgaHwang committed:
test: StudentManager 동작을 위한 테스트 파일 추가 반영

volgaHwang committed:
test: StudentManager에서 학생 추가 / 삭제에 관한 테스트 케이스를 반영합니다.

- StudentManager에서 학생 추가 / 삭제에 관련된 테스트케이스를 반영했습니다.
    - 학생 추가 시, 중복이 될 경우와, 중복이 되지 않는 경우에 대한 테스트케이스를 생성했습니다.
    - 학생 삭제 시, 중복인 이름을 삭제 할 경우에 대해서 테스트케이스를 생성했습니다.

volgaHwang committed:
feat: StudentManager에서 보유하고있는 list의 Count를 리턴해주는 Computed Property를 반영합니다.

volgaHwang committed:
chore: Student가 소유할 Grade 객체에 대한 테스트 파일을 추가합니다.

volgaHwang committed:
feat: CouresManager를 추가하여 Student가 소유하게 했고, 이를 통해 학생 성적을 관리하는 기능을 추가하였습니다.

- Student의 관리를 위해서 Class로 변경 했습니다.
    - StudentManager에서 Set<Student>를 순회하는 경우에 Struct를 사용 할 경우 값이 복사가 되어 따로 주소값을 통한 Indicating이 불가능한 상황이고 따라서 원본에 필요한 메소드를 출력하지 못하여 Class로 전환했습니다.
- CourseManager 구조체에선 dictionary 구조로 [과목명: 성적Enumerate] 구조를 사용하여 과목별 성적을 관리합니다.
    - 관리되는 딕셔너리의 list를 통해 CourseManager는 상기 목표였던 성적 추가 및 변경 그리고 삭제 기능까지 추가하였습니다.
- 추가적으로 리팩토링 한 부분이 있는데, isDuplicate는 의미가 모호하여 isExist와 같은 네이밍으로 변경을 진행한 부분이 있습니다.

volgaHwang committed:
feat: CourseManager에 평점 기능을 추가했습니다.
- 총점을 볼 수 있도록 기능 추가를 했습니다

Commit:
chore: 파일 분리

Commit:
refactor: 형식 개선, 고차함수 사용등 간단한 리팩토링

wonhee009 commented on GradeManager/Student.swift:
```diff
@@ -0,0 +1,42 @@
+
+class Student: Hashable {
+    private var name: String
+    private var courseManager: CourseManager
```
Student가 courseManager를 갖고 있는 부분이 조금 부자연스러운거 같은데요.
`CourseManager`가 InfoMessage나 Grade에 대해 가지고 있는거로 보입니다.
Student가 courseManger를 갖으면서 모든 student 객체는 infoMesage, Grade enum에 대한 타입을 갖고 있게 될거 같습니다.
이런 타입이 모든 student 객체가 갖는게 의도하신 방향일까요?

wonhee009 commented on GradeManager/InputManager.swift:
```diff
@@ -0,0 +1,91 @@
+}
+
+struct InputManager {
+    func viewAverageScore() -> Result<String, InputError> {
+        guard let studentName = readLine() else {
+            return .failure(.invalidUserInput)
+        }
+        return .success(studentName)
+    }
```
메서드명은 평점을 보여줘야할거 같은데 학생 이름을 반환하고 있는거 같습니다.
메서드명과 역할이 동일하도록 해주세요.

wonhee009 commented on GradeManager/InputManager.swift:
```diff
@@ -0,0 +1,91 @@
+        return .success(studentName)
+    }
+    
+    func addStudentGradeInfo() -> Result<(name: String, subject: String, grade: String), InputError> {
```
이후에 info에 학생 전화번호가 추가되면 `(name: _, subject: _, grade: _, phoneNumber: _)` 이런식으로 무한히 확장되는 걸까요?
`Student`나 `GradeInfo` 같이 모델을 추가하는건 어떤가요?

shinhee-rebecca replied:
확장성을 고려했을 때 name: String, subject: String, grade: String 보다는 student: Student를 Result에 포함하는 것이 적절해보입니다. 감사합니다!

wonhee009 commented on GradeManager/StudentManager.swift:
```diff
@@ -0,0 +1,98 @@
+    func viewAverageScore(name: String)
+}
+
+struct DefaultStudentManager: StudentManager {
```
이전 코멘트에도 남겼지만 Default는 어떤 의미인가요? 다른 StudentManger가 있는걸까요?
(https://github.com/2023newkrew/ios-grade-manager/pull/1#discussion_r1058001361)

shinhee-rebecca replied:
StudentManger 라는 이름의 protocol이 존재하여 struct의 이름에 Default를 붙였습니다. struct에 default를 붙이는 것 말고 protocol과 struct의 네이밍 방식에는 어떤 것이 있을까요?

wonhee009 commented on GradeManager/StudentManager.swift:
```diff
@@ -0,0 +1,98 @@
+    }
+    
+    func isExist(name: String) -> Bool {
+        return list.contains(Student(name: name))
```
입력받은 name으로 Student 객체를 만들고 새로 만든 객체를 통해 contains 여부를 확인하는거로 이해했습니다.
name이 동일하다면 true를 반환하는거로 이해했습니다. Student 객체를 만들지 않고 확인할 수 없을까요?
contains를 수행하기 위해 새로운 객체를 만드는게 어색한거 같습니다.

wonhee009 replied:
이 메서드는 StudentManager 내부에서만 사용하나요? 그렇다면 private으로 외부 접근을 막을 필요는 없을까요?

shinhee-rebecca replied:
Student 객체를 만들지 않고도 확인할 수 있을 것 같습니다. contains를 수행하기 위해 객체를 만드는건 저도 어색하게 보입니다. private 접근 지정을 하는 것이 좋은 것 같습니다.

wonhee009 commented on GradeManager/StudentManager.swift:
```diff
@@ -0,0 +1,98 @@
+        return list.contains(Student(name: name))
+    }
+    
+    func studentToBeDeleted(name: String) -> Student? {
```
`deleteStudent()`에서 받은 이름이 삭제 가능한 학생의 이름인지 확인하는 메서드로 이해했습니다.
`willDeleteStudent`정도의 네이밍은 어떤가요?

wonhee009 replied:
또한 해당 메서드 StudentManager 내에서만 사용될거 같은데요. private으로 외부 접근을 막을 필요는 없을까요?

shinhee-rebecca replied:
위 메서드는 삭제할 학생을 찾아서 반환하는 메서드입니다. (학생을 찾지 못하면 nil을 반환하게 됩니다.) 해당 메서드를 private으로 막는것이 더 적절할 것 같습니다.

wonhee009 reviewed (changes requested):
수고하셨습니다.
코멘트 남겨두었으니 확인 부탁 드립니다.

- Student를 구조체로 유지하고 싶으신 이유가 따로 있으신걸까요? mutating을 이용해 내부를 변경할 수는 있습니다만, 구조체와 클래스의 차이점에 대해 좀 더 생각해보면 좋을거 같습니다. 이전 코멘트에서 적어주셔듯이 어떤 경우에 구조체/클래스를 사용하도록 권장하는지에 대해 생각해보면 좋을거 같습니다. (저는 구조체가 immutable한 속성을 갖고 있는만큼 mutating한 행위를 할 필요가 있다면 class로 정의하는게 맞다고 생각합니다.)
